tutorial/tests/test_functions_advanced.py

```diff
@@ -0,0 +1,259 @@
+import pathlib
+import time
+import typing as t
+from operator import sub
+from string import ascii_lowercase as lowercase
+from string import ascii_uppercase as uppercase
+from string import digits, punctuation
+
+import pytest
+
+
+#
+# Exercise: Password checker factory
+#
+def reference_password_checker_factory(
+    min_up: int, min_low: int, min_pun: int, min_dig: int
+) -> t.Callable:
+    """Password checker factory"""
+
+    def password_checker(password: str) -> tuple[bool, dict]:
+        """Password checker function"""
+
+        # Counts the number of chars for each class in a password
+        counts = [
+            sum(1 for char in password if char in _class)
+            for _class in (uppercase, lowercase, punctuation, digits)
+        ]
+
+        # Compare with requirements and calculate the differences
+        diffs = [
+            sub(*pair) for pair in zip(counts, (min_up, min_low, min_pun, min_dig))
+        ]
+
+        result = dict(zip(("uppercase", "lowercase", "punctuation", "digits"), diffs))
+
+        return all(diff >= 0 for diff in diffs), result
+
+    return password_checker
+
+
+def test_password_checker_factory_no_min_no_pw(function_to_test: t.Callable):
```

Comment:
I think all the tests could be wrapped up in something like this, what do you think?
```python
@pytest.mark.parametrize(
    "args, pwd",
    [
        ((0, 0, 0, 0), ""),
        ((0, 0, 0, 0), "ABCDefgh!@#$1234"),
    ]
)
def test_password_checker_factory(args, pwd, function_to_test):
    ref = reference_password_checker_factory(*args)
    student_solution = function_to_test(*args)
    assert ref(pwd) == student_solution(pwd)

```

I am a bit surprised that you defined almost all of them explicitly.

Reply:
Agree, that's more compact, but I wanted tests with different names so that it's cleared what pytest is testing. Otherwise, you see a single test_name with appended the values of the parametrization currently tested.